src/custom_request_schedule_manager.h

```diff
@@ -0,0 +1,140 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include "load_manager.h"
+#include "request_rate_manager.h"
+
+namespace triton { namespace perfanalyzer {
+
+//==============================================================================
+/// CustomRequestScheduleManager is a helper class to send inference requests to
+/// inference server in accordance with the schedule set by the user.
+///
+/// Detail:
+/// An instance of this load manager will be created at the beginning of the
+/// perf analyzer and it will be used to schedule to send requests at that
+/// particular second defined by the user. The particular seconds at which a
+/// request should be sent can be set by the user using the `schedule` option.
+/// For example, if the `schedule` is set to 1,2,4,5,6.5,
+/// CustomRequestScheduleManager sends request at 1st second, 2nd second, 4th
+/// second and so on.
+///
+
+class CustomRequestScheduleManager : public RequestRateManager {
+ public:
+  ~CustomRequestScheduleManager() = default;
+
+  /// Creates an object of CustomRequestScheduleManager
+  /// \param async Whether to use asynchronous or synchronous API for infer
+  /// request \param streaming Whether to use gRPC streaming API for infer
+  /// request \param measurement_window_ms The time window for measurements
+  /// \param max_trials The maximum number of windows that will be measured
+  /// \param schedule The vector containing the schedule for requests
+  /// \param batch_size The batch size used for each request
+  /// \param max_threads The maximum number of working threads to be spawned
+  /// \param num_of_sequences The number of concurrent sequences to maintain on
+  /// the server \param shared_memory_type The type of shared memory to use for
+  /// inputs \param output_shm_size The size of the shared memory to allocate
+  /// for the output \param serial_sequences Enable serial sequence mode \param
+  /// parser The ModelParser object to get the model details \param factory The
+  /// ClientBackendFactory object used to create client to the server \param
+  /// manager Returns a new CustomRequestScheduleManager object \param
+  /// request_parameters Custom request parameters to send to the server \return
+  /// cb::Error object indicating success or failure
+  static cb::Error Create(
```

Comment:
To clarify a bit my stance: I am not against smart pointers in general, just that `shared_ptr` are usually `unique_ptr` in disguise. They are non-committal about the actual ownership of the object, whereas `unique_ptr` has a clearer definition of it.

Reply:
The shared_ptr is for client backend factory and I have no idea how much of an effort it is to modify everything to unique_ptr. I can create a ticket for that effort.
Coming to the constructor call, I agree with all your points. As far as I understand, your suggestion is to pass params object instead of the individual arguments. It makes sense to me. Again, I'm not sure if I'm going to break something by modifying the API. Let me know if I should add a ticket to investigate and modify the all the related APIs because as you said, it might look awkward if we modify for some of them and it's not propagated until some time after.

Reply:
Right, let's just try and plumb the parameters object in this one instance, and keep in mind we should do this for newer code. If we have time, we should go back and clean older API incrementally. You're right: let's not break existing code right now, but let's avoid introducing new code which follows this older pattern.

For the unique_ptr vs shared_ptr, this was more of a general remark that we should keep in mind when adding new shared_ptr in the codebase. At some point we should review the existing usage and clean it up, but not in this one PR.